tensorcircuit/results/readout_mitigation.py

```diff
@@ -652,20 +737,43 @@ def expectation(
         :param diagoal_op: shape [n, 2], explicitly indicate the diagonal op on each qubit
             eg. [1, -1] for z [1, 1] for I, etc.
         :type diagoal_op: Tensor
+        :param positional_logical_mapping: positional_logical_mapping, defaults to None.
+        :type positional_logical_mapping: Optional[Dict[int, int]], optional
+        :param logical_physical_mapping: logical_physical_mapping, defaults to None
+        :type logical_physical_mapping: Optional[Dict[int, int]], optional
         :param method: readout mitigation method, defaults to "constrained_least_square"
         :type method: str, optional
         :return: expectation value after readout error mitigation
         :rtype: float
         """
         # https://arxiv.org/pdf/2006.14044.pdf
 
-        if z is None and diagonal_op is None:
-            raise ValueError("One of `z` and `diagonal_op` must be set")
+        # count[0,1,2], logical[3,4,5], physic[6,7,8], z=[4,5](logical)
+        # z1=[1,2](position), z=[7,8] (physic)
+        # diagonal_op [i6 z7 z8 ]
+
         n = len(list(counts.keys())[0])
 
+        if logical_physical_mapping is None:
+            if positional_logical_mapping is None:
+                logical_qubits = list(range(n))
+            else:
+                logical_qubits = list(positional_logical_mapping.values())
+            physical_qubits = logical_qubits
+        else:
+            logical_qubits = list(logical_physical_mapping.keys())
+            physical_qubits = list(logical_physical_mapping.values())
```

Comment:
``physical_qubits = [logical_physical_mapping[positional_logical_mapping[pq]] for pq in range(n)]``